<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <title>Chapter-10-Item-73-Throw-exceptions-appropriate-to-the-abstraction</title>
  <style>
    html {
      color: #1a1a1a;
      background-color: #fdfdfd;
    }
    body {
      margin: 0 auto;
      max-width: 36em;
      padding-left: 50px;
      padding-right: 50px;
      padding-top: 50px;
      padding-bottom: 50px;
      hyphens: auto;
      overflow-wrap: break-word;
      text-rendering: optimizeLegibility;
      font-kerning: normal;
    }
    @media (max-width: 600px) {
      body {
        font-size: 0.9em;
        padding: 12px;
      }
      h1 {
        font-size: 1.8em;
      }
    }
    @media print {
      html {
        background-color: white;
      }
      body {
        background-color: transparent;
        color: black;
        font-size: 12pt;
      }
      p, h2, h3 {
        orphans: 3;
        widows: 3;
      }
      h2, h3, h4 {
        page-break-after: avoid;
      }
    }
    p {
      margin: 1em 0;
    }
    a {
      color: #1a1a1a;
    }
    a:visited {
      color: #1a1a1a;
    }
    img {
      max-width: 100%;
    }
    svg {
      height: auto;
      max-width: 100%;
    }
    h1, h2, h3, h4, h5, h6 {
      margin-top: 1.4em;
    }
    h5, h6 {
      font-size: 1em;
      font-style: italic;
    }
    h6 {
      font-weight: normal;
    }
    ol, ul {
      padding-left: 1.7em;
      margin-top: 1em;
    }
    li > ol, li > ul {
      margin-top: 0;
    }
    blockquote {
      margin: 1em 0 1em 1.7em;
      padding-left: 1em;
      border-left: 2px solid #e6e6e6;
      color: #606060;
    }
    code {
      font-family: Menlo, Monaco, Consolas, 'Lucida Console', monospace;
      font-size: 85%;
      margin: 0;
      hyphens: manual;
    }
    pre {
      margin: 1em 0;
      overflow: auto;
    }
    pre code {
      padding: 0;
      overflow: visible;
      overflow-wrap: normal;
    }
    .sourceCode {
     background-color: transparent;
     overflow: visible;
    }
    hr {
      background-color: #1a1a1a;
      border: none;
      height: 1px;
      margin: 1em 0;
    }
    table {
      margin: 1em 0;
      border-collapse: collapse;
      width: 100%;
      overflow-x: auto;
      display: block;
      font-variant-numeric: lining-nums tabular-nums;
    }
    table caption {
      margin-bottom: 0.75em;
    }
    tbody {
      margin-top: 0.5em;
      border-top: 1px solid #1a1a1a;
      border-bottom: 1px solid #1a1a1a;
    }
    th {
      border-top: 1px solid #1a1a1a;
      padding: 0.25em 0.5em 0.25em 0.5em;
    }
    td {
      padding: 0.125em 0.5em 0.25em 0.5em;
    }
    header {
      margin-bottom: 4em;
      text-align: center;
    }
    #TOC li {
      list-style: none;
    }
    #TOC ul {
      padding-left: 1.3em;
    }
    #TOC > ul {
      padding-left: 0;
    }
    #TOC a:not(:hover) {
      text-decoration: none;
    }
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    /* The extra [class] is a hack that increases specificity enough to
       override a similar rule in reveal.js */
    ul.task-list[class]{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      font-size: inherit;
      width: 0.8em;
      margin: 0 0.8em 0.2em -1.6em;
      vertical-align: middle;
    }
    .display.math{display: block; text-align: center; margin: 0.5rem auto;}
  </style>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h2 id="chapter-10.-exceptions异常">Chapter 10. Exceptions（异常）</h2>
<h3
id="item-73-throw-exceptions-appropriate-to-the-abstraction抛出能用抽象解释的异常">Item
73: Throw exceptions appropriate to the
abstraction（抛出能用抽象解释的异常）</h3>
<p>It is disconcerting when a method throws an exception that has no
apparent connection to the task that it performs. This often happens
when a method propagates an exception thrown by a lower-level
abstraction. Not only is it disconcerting, but it pollutes the API of
the higher layer with implementation details. If the implementation of
the higher layer changes in a later release, the exceptions it throws
will change too, potentially breaking existing client programs.</p>
<p>当一个方法抛出一个与它所执行的任务没有明显关联的异常时，这是令人不安的。这种情况经常发生在由方法传播自低层抽象抛出的异常。它不仅令人不安，而且让实现细节污染了上层的
API。如果高层实现在以后的版本中发生变化，那么它抛出的异常也会发生变化，可能会破坏现有的客户端程序。</p>
<p>To avoid this problem, <strong>higher layers should catch lower-level
exceptions and, in their place, throw exceptions that can be explained
in terms of the higher-level abstraction.</strong> This idiom is known
as exception translation:</p>
<p>为了避免这个问题，<strong>高层应该捕获低层异常，并确保抛出的异常可以用高层抽象解释。</strong>
这个习惯用法称为异常转换：</p>
<pre><code>// Exception Translation
try {
    ... // Use lower-level abstraction to do our bidding
} catch (LowerLevelException e) {
    throw new HigherLevelException(...);
}</code></pre>
<p>Here is an example of exception translation taken from the
AbstractSequentialList class, which is a skeletal implementation (Item
20) of the List interface. In this example, exception translation is
mandated by the specification of the get method in the
<code>List&lt;E&gt;</code> interface:</p>
<p>下面是来自 AbstractSequentialList 类的异常转换示例，该类是 List
接口的一个框架实现（<a
href="Chapter-4-Item-20-Prefer-interfaces-to-abstract-classes.html">Item-20</a>）。在本例中，异常转换是由
<code>List&lt;E&gt;</code> 接口中的 get 方法规范强制执行的：</p>
<pre><code>/**
* Returns the element at the specified position in this list.
* @throws IndexOutOfBoundsException if the index is out of range
* ({@code index &lt; 0 || index &gt;= size()}).
*/
public E get(int index) {
    ListIterator&lt;E&gt; i = listIterator(index);
    try {
        return i.next();
    }
    catch (NoSuchElementException e) {
        throw new IndexOutOfBoundsException(&quot;Index: &quot; + index);
    }
}</code></pre>
<p>A special form of exception translation called exception chaining is
called for in cases where the lower-level exception might be helpful to
someone debugging the problem that caused the higher-level exception.
The lower-level exception (the cause) is passed to the higher-level
exception, which provides an accessor method (Throwable’s getCause
method) to retrieve the lower-level exception:</p>
<p>如果低层异常可能有助于调试高层异常的问题，则需要一种称为链式异常的特殊异常转换形式。低层异常（作为原因）传递给高层异常，高层异常提供一个访问器方法（Throwable
的 getCause 方法）来检索低层异常：</p>
<pre><code>// Exception Chaining
try {
    ... // Use lower-level abstraction to do our bidding
}
catch (LowerLevelException cause) {
    throw new HigherLevelException(cause);
}</code></pre>
<p>The higher-level exception’s constructor passes the cause to a
chaining-aware superclass constructor, so it is ultimately passed to one
of Throwable’s chaining-aware constructors, such as
Throwable(Throwable):</p>
<p>高层异常的构造函数将原因传递给能够接收链式异常的超类构造函数，因此它最终被传递给
Throwable 的一个接收链式异常的构造函数，比如
<code>Throwable(Throwable)</code>：</p>
<pre><code>// Exception with chaining-aware constructor
class HigherLevelException extends Exception {
    HigherLevelException(Throwable cause) {
        super(cause);
    }
}</code></pre>
<p>Most standard exceptions have chaining-aware constructors. For
exceptions that don’t, you can set the cause using Throwable’s initCause
method. Not only does exception chaining let you access the cause
programmatically (with getCause), but it integrates the cause’s stack
trace into that of the higher-level exception.</p>
<p>大多数标准异常都有接收链式异常的构造函数。对于不支持链式异常的异常，可以使用
Throwable 的 initCause
方法设置原因。异常链接不仅允许你以编程方式访问原因（使用
getCause），而且还将原因的堆栈跟踪集成到更高层异常的堆栈跟踪中。</p>
<p><strong>While exception translation is superior to mindless
propagation of exceptions from lower layers, it should not be
overused.</strong> Where possible, the best way to deal with exceptions
from lower layers is to avoid them, by ensuring that lower-level methods
succeed. Sometimes you can do this by checking the validity of the
higher-level method’s parameters before passing them on to lower
layers.</p>
<p>虽然异常转换优于底层异常的盲目传播，但它不应该被过度使用。在可能的情况下，处理低层异常的最佳方法是确保低层方法避免异常。有时，你可以在将高层方法的参数传递到低层之前检查它们的有效性。</p>
<p>If it is impossible to prevent exceptions from lower layers, the next
best thing is to have the higher layer silently work around these
exceptions, insulating the caller of the higher-level method from
lower-level problems. Under these circumstances, it may be appropriate
to log the exception using some appropriate logging facility such as
java.util.logging. This allows programmers to investigate the problem,
while insulating client code and the users from it.</p>
<p>如果不可能从低层防止异常，那么下一个最好的方法就是让高层静默处理这些异常，使较高层方法的调用者免受低层问题的影响。在这种情况下，可以使用一些适当的日志工具（如
<code>java.util.logging</code>）来记录异常。这允许程序员研究问题，同时将客户端代码和用户与之隔离。</p>
<p>In summary, if it isn’t feasible to prevent or to handle exceptions
from lower layers, use exception translation, unless the lower-level
method happens to guarantee that all of its exceptions are appropriate
to the higher level. Chaining provides the best of both worlds: it
allows you to throw an appropriate higherlevel exception, while
capturing the underlying cause for failure analysis (Item 75).</p>
<p>总之，如果无法防止或处理来自低层的异常，则使用异常转换，但要保证低层方法的所有异常都适用于较高层。链式异常提供了兼顾两方面的最佳服务：允许抛出适当的高层异常，同时捕获并分析失败的潜在原因（<a
href="Chapter-10-Item-75-Include-failure-capture-information-in-detail-messages.html">Item-75</a>）。</p>
<hr />
<p><strong><a href="Chapter-10-Introduction.html">Back to
contents of the chapter（返回章节目录）</a></strong> - <strong>Previous
Item（上一条目）：<a
href="Chapter-10-Item-72-Favor-the-use-of-standard-exceptions.html">Item
72: Favor the use of standard
exceptions（鼓励复用标准异常）</a></strong> - <strong>Next
Item（下一条目）：<a
href="Chapter-10-Item-74-Document-all-exceptions-thrown-by-each-method.html">Item
74: Document all exceptions thrown by each
method（为每个方法记录会抛出的所有异常）</a></strong></p>
</body>
</html>
